utilities/composite.1
.ad l
.nh
.TH composite 1 "10 January 1993" "ImageMagick"
.SH NAME
composite - composite images to create new images.
.SH SYNOPSIS
.B "composite"
[ \fIoptions\fP ... ] \fIimage\fP \fIcomposite\fP [ \fImask\fP ] \fIcomposited\fP
.SH DESCRIPTION
\fBcomposite\fP composite images to create new images.
.SH EXAMPLES

To composite a image of a cockatoo with a perch, use

.nf
     composite cockatoo.miff perch.ras composite.miff
.fi

To compute the difference between images in a series, use

.nf
     composite -compose difference series.1 series.2 difference.miff
.fi

To composite a image of a cockatoo with a perch starting at location (100,150),
use

.nf
     composite -geometry +100+150 cockatoo.miff perch.ras composite.miff
.fi

To tile a logo across your image of a cockatoo, use

.nf
     convert +shade 30x60 cockatoo.miff mask.miff
     composite -compose bumpmap -tile logo.gif cockatoo.miff mask.miff composite.miff

.fi
To composite a red, green, and blue color plane into a single composite image,
try

.nf
     composite -compose CopyGreen red.png green.png red-green.png
     composite -compose CopyBlue red-green.png blue.png composite.png

.fi
.SH OPTIONS
.TP
.B "-cache \fIthreshold\fP"
megabytes of memory available to the pixel cache.

Image pixels are stored in memory until 80 megabytes of memory
have been consumed.  Subsequent pixel operations are cached on disk.
Operations to memory are significantly faster but if your computer does not
have a sufficient amount of free memory you may want to adjust this threshold
value.
.TP
.B "-colors \fIvalue\fP"
preferred number of colors in the image.

The actual number of colors in the image may be less than your request,
but never more.  Note, this is a color reduction option.  Images with
less unique colors than specified with this option will have any duplicate
or unused colors removed.
Refer to \fBquantize(9)\fP for more details.

Note, options \fB-dither\fP, \fB-colorspace\fP, and \fB-treedepth\fP affect
the color reduction algorithm.
.TP
.B "-colorspace \fIvalue\fP"
the type of colorspace: \fBGRAY\fP, \fBOHTA\fP, \fBRGB\fP,
\fBTransparent\fP, \fBXYZ\fP, \fBYCbCr\fP, \fBYIQ\fP, \fBYPbPr\fP,
\fBYUV\fP, or \fBCMYK\fP.

Color reduction, by default, takes place in the RGB color space.
Empirical evidence suggests that distances in color spaces such as YUV
or YIQ correspond to perceptual color differences more closely
than do distances in RGB space.  These color spaces may give better
results when color reducing an image.  Refer to \fBquantize(9)\fP for
more details.

The \fBTransparent\fP color space behaves uniquely in that it preserves
the matte channel of the image if it exists.

The \fB-colors\fP or \fB-monochrome\fP option is required for this option
to take effect.
.TP
.B "-comment \fIstring\fP"
annotate an image with a comment.

By default, each image is commented with its file name.  Use this
option to assign a specific comment to the image.  Optionally you can
include the image filename, type, width, height, or other image
attributes by embedding special format characters:

.nf
    %b   file size
    %c   comment
    %d   directory
    %e   filename extention
    %f   filename
    %h   height
    %i   input filename
    %k   number of unique colors
    %l   label
    %m   magick
    %n   number of scenes
    %o   output filename
    %p   page number
    %q   quantum depth
    %s   scene number
    %t   top of filename
    %u   unique temporary filename
    %w   width
    %x   x resolution
    %y   y resolution
    \\n   newline
    \\r   carriage return
.fi

For example,

.nf
     -comment "%m:%f %wx%h"
.fi

produces an image comment of \fBMIFF:bird.miff 512x480\fP for an image
titled \fBbird.miff\fP and whose width is 512 and height is 480.

If the first character of \fIstring\fP is \fB@\fP, the image comment is read
from a file titled by the remaining characters in the string.
.TP
.B "-compose \fIoperator\fP"
the type of image composition.

By default, each of the composite image pixels are replaced by the
corresponding image tile pixel. You can choose an alternate composite
operation:

.nf
    Over
    In
    Out
    Atop
    Xor
    Plus
    Minus
    Add
    Subtract
    Difference
    Multiply
    Bumpmap
    Copy
    CopyRed
    CopyGreen
    CopyBlue
    CopyOpacity
.fi

How each operator behaves is described below.
.TP
.B over
The result will be the union of the two image shapes, with
\fIcomposite image\fP obscuring \fIimage\fP in the region of overlap.
.TP
.B In
The result is simply \fIcomposite image\fP cut by the shape of \fIimage\fP.
None of the image data of \fIimage\fP will be in the result.
.TP
.B Out
The resulting image is \fIcomposite image\fP with the shape of \fIimage\fP cut
out.
.TP
.B Atop
The result is the same shape as image \fIimage\fP, with \fIcomposite image\fP
obscuring \fIimage\fP where the image shapes overlap.  Note this differs from
\fBover\fP because the portion of \fIcomposite image\fP outside \fIimage\fP's
shape does not appear in the result.
.TP
.B Xor
The result is the image data from both \fIcomposite image\fP and \fIimage\fP
that is outside the overlap region.  The overlap region will be blank.
.TP
.B Plus
The result is just the sum of the image data.  Output values are
cropped to 255 (no overflow).  This operation is independent of the
matte channels.
.TP
.B Minus
The result of \fIcomposite image\fP \- \fIimage\fP, with underflow cropped to
zero.  The matte channel is ignored (set to 255, full coverage).
.TP
.B Add
The result of \fIcomposite image\fP + \fIimage\fP, with overflow wrapping
around (\fImod\fP 256).
.TP
.B Subtract
The result of \fIcomposite image\fP - \fIimage\fP, with underflow wrapping
around (\fImod\fP 256).  The \fBadd\fP and \fBsubtract\fP operators can be
used to perform reversible transformations.
.TP
.B Difference
The result of abs(\fIcomposite image\fP \- \fIimage\fP).  This is useful for
comparing two very similar images.
.TP
.B Multipy
The result of \fIcomposite image\fP \* \fIimage\fP.  This is useful for
the creation of drop-shadows.
.TP
.B Bumpmap
The result \fIimage\fP shaded by \fIcomposite image\fP.
.TP
.B Copy
The resulting image is \fIimage\fP replaced with \fIcomposite image\fP.  Here
the matte information is ignored.
.TP
.B CopyRed
The resulting image is the red layer in \fIimage\fP replaced with the red
layer in \fIcomposite image\fP.  The other layers are copied untouched.
.TP
.B CopyGreen
The resulting image is the green layer in \fIimage\fP replaced with the green
layer in \fIcomposite image\fP.  The other layers are copied untouched.
.TP
.B CopyBlue
The resulting image is the blue layer in \fIimage\fP replaced with the blue
layer in \fIcomposite image\fP.  The other layers are copied untouched.
.TP
.B CopyOpacity
The resulting image is the matte layer in \fIimage\fP replaced with the matte
layer in \fIcomposite image\fP.  The other layers are copied untouched.

The image compositor requires a matte, or alpha channel in the image
for some operations.  This extra channel usually defines a mask which
represents a sort of a cookie-cutter for the image.  This is the case
when matte is 255 (full coverage) for pixels inside the shape, zero
outside, and between zero and 255 on the boundary.  For certain
operations, if \fIimage\fP does not have a matte channel, it is initialized
with 0 for any pixel matching in color to pixel location (0,0), otherwise
255 (to work properly \fBborderwidth\fP must be 0).
.TP
.B "-compress \fItype\fP"
the type of image compression: \fINone\fP, \fIBZip\fP, \fIFax\fP, \fIGroup4\fP,
\fIJPEG\fP, \fILZW\fP, \fIRLE\fP, or \fIZip\fP.

Specify \fB\+compress\fP to store the binary image in an uncompressed format.
The default is the compression type of the specified image file.
.TP
.B "-density \fI<width>x<height>\fP
vertical and horizontal resolution in pixels of the image.

This option specifies an image density when decoding a Postscript or Portable
Document page.  The default is 72 pixels per inch in the horizontal and
vertical direction.  This option is used in concert with \fB-page\fP.
.TP
.B "-depth \fIvalue\fP"
depth of the image.  This is the number of bits in a pixel.  The only
acceptable values are 8 or 16.
.TP
.B "-displace \fI<horizontal scale>x<vertical scale>\fP"
shift image pixels as defined by a displacement map.

With this option, \fIcomposite image\fP is used as a displacement map.  Black,
within the displacement map, is a maximum positive displacement.  White is a
maximum negative displacement and middle gray is neutral.  The displacement
is scaled to determine the pixel shift.  By default, the displacement applies
in both the horizontal and vertical directions.  However, if you specify
\fImask\fP, \fIcomposite image\fP is the horizontal X displacement and
\fImask\fP the vertical Y displacement.
.TP
.B "-display \fIhost:display[.screen]\fP"
specifies the X server to contact; see \fBX(1)\fP.
.TP
.B "-dispose \fImethod\fP"
GIF disposal method.

Here are the valid methods:

.nf
     0     No disposal specified.
     1     Do not dispose between frames.
     2     Overwrite frame with background color from header.
     3     Overwrite with previous frame.
.fi
.TP
.B "-dissolve \fIvalue\fP"
dissolve the two images a given percent.
.TP
.B "-dither"
apply Floyd/Steinberg error diffusion to the image.

The basic strategy of dithering is to trade intensity resolution for
spatial resolution by averaging the intensities of several neighboring
pixels.  Images which suffer from severe contouring when reducing colors
can be improved with this option.

The \fB-colors\fP or \fB-monochrome\fP option is required for this option
to take effect.

Use \fB+dither\fP to render Postscript without text or graphic aliasing.
.TP
.B "-font \fIname\fP"
This option specifies the font to be used  for displaying normal text.

If the font is a fully qualified X server font name, the font is obtained
from an X server (e.g. -*-helvetica-medium-r-*-*-12-*-*-*-*-*-iso8859-*).  To
use a TrueType font, precede the TrueType filename with a \fB@\fP (e.g.
@times.ttf).  Otherwise, specify a Postscript, X11, or TrueType font (e.g. helvetica).
.TP
.B "-geometry \fI{\+-}<x offset>{\+-}<y offset>{%}{@}{!}{<}{>}\fP"
location of the composite image.

By default the images are composited relative to the image gravity
(see \fB-gravity\fP).  Use \fI<x offset>\fP and
\fI<y offset>\fP to specify a particular location to composite the images.
.TP
.B "-gravity \fItype\fP"
direction image gravitates to within the composite: NorthWest, North,
NorthEast, West, Center, East, SouthWest, South, SouthEast.
See \fBX(1)\fP for details about the gravity specification.

The image may not fill the composite completely (see
\fB-geometry\fP).  The direction you choose specifies where to
position the image within the composite.  For example \fICenter\fP gravity
forces the image to be centered within the composite.  
By default, the image gravity is \fINorthWest\fP.
.TP
.B "-interlace \fItype\fP"
the type of interlacing scheme: \fBNone\fP, \fBLine\fP, \fBPlane\fP, or
\fBPartition\fP.  The default is \fBNone\fP.

This option is used to specify the type of interlacing scheme for raw
image formats such as \fBRGB\fP or \fBYUV\fP.  \fBNo\fP means do not
interlace (RGBRGBRGBRGBRGBRGB...), \fBLine\fP uses scanline
interlacing (RRR...GGG...BBB...RRR...GGG...BBB...), and \fBPlane\fP uses
plane interlacing (RRRRRR...GGGGGG...BBBBBB...).  \fBPartition\fP is like
plane except the different planes are saved to individual files (e.g.
image.R, image.G, and image.B).

Use \fBLine\fP, or \fBPlane\fP to create an interlaced GIF or progressive
JPEG image.
.TP
.B "-label \fIname\fP"
assign a label to an image.

Use this option to assign a specific label to the image.  Optionally
you can include the image filename, type, width, height, or other image
attributes by embedding special format characters.  See
\fB-comment\fP for details.

For example,

.nf
     -label "%m:%f %wx%h"
.fi

produces an image label of \fBMIFF:bird.miff 512x480\fP for an image
titled \fBbird.miff\fP and whose width is 512 and height is 480.

If the first character of \fIstring\fP is \fB@\fP, the image label is read
from a file titled by the remaining characters in the string.

When converting to Postscript, use this option to specify a header string
to print above the image. Specify the label font with \fB-font\fP.
.TP
.B "-matte"
store matte channel if the image has one otherwise create an opaque one.
.TP
.B "-monochrome"
transform the image to black and white.
.TP
.B "-negate"
replace every pixel with its complementary color (white becomes black, yellow
becomes blue, etc.).

The red, green, and blue intensities of an image are negated.  Use
\fB+negate\fP to only negate the grayscale pixels of the image.
.TP
.B "-page \fI<width>x<height>{\+-}<x offset>{\+-}<y offset>{%}{!}{<}{>}\fP"
preferred size and location of an image canvas.

Use this option to specify the dimensions of the Postscript page in
dots per inch or a TEXT page in pixels.  The choices for a Postscript page are:

.nf
       11x17         792  1224 
       Ledger       1224   792    
       Legal         612  1008
       Letter        612   792
       LetterSmall   612   792
       ArchE        2592  3456
       ArchD        1728  2592
       ArchC        1296  1728
       ArchB         864  1296
       ArchA         648   864
       A0           2380  3368
       A1           1684  2380
       A2           1190  1684
       A3            842  1190
       A4            595   842
       A4Small       595   842
       A5            421   595
       A6            297   421
       A7            210   297
       A8            148   210
       A9            105   148
       A10            74   105
       B0           2836  4008
       B1           2004  2836
       B2           1418  2004
       B3           1002  1418
       B4            709  1002
       B5            501   709
       C0           2600  3677
       C1           1837  2600
       C2           1298  1837
       C3            918  1298
       C4            649   918
       C5            459   649
       C6            323   459
       Flsa          612   936 
       Flse          612   936
       HalfLetter    396   612
.fi

For convenience you can specify the page size by media (e.g.
A4, Ledger, etc.).  Otherwise, \fB-page\fP behaves much like \fB-geometry\fP
(e.g. -page letter+43+43>).

To position a GIF image, use -page \fI{\+-}<x offset>{\+-}<y offset>\fP
(e.g. -page +100+200).

For a Postscript page, the image is sized as in \fB-geometry\fP and
positioned relative to the lower left hand corner of the page by
\fI{\+-}<x offset>{\+-}<y offset>\fP.  Use -page 612x792>, for example,
to center the image within the page.  If the image size exceeds the
Postscript page, it is reduced to fit the page.

The default page dimensions for a TEXT image is 612x792.

This option is used in concert with \fB-density\fP.
.TP
.B "-quality \fIvalue\fP"
JPEG/MIFF/PNG compression level.

For the JPEG image format, quality is 0 (worst) to 100 (best).  The default
quality is 75.

Quality for the MIFF and PNG image format sets the amount of image compression
(quality / 10) and filter-type (quality % 10).  Compression quality
values range from 0 (worst) to 100 (best).  If filter-type is 4
or less, the specified filter-type is used for all scanlines:

.nf
    0: none
    1: sub
    2: up
    3: average
    4: Paeth
.fi

If filter-type is 5, adaptive filtering is used when quality
is greater than 50 and the image does not have a color map,
otherwise no filtering is used.

If filter-type is 6 or more, adaptive filtering with
\fIminimum-sum-of-absolute-values\fP is used.

The default is quality is 75.  Which means nearly the best compression
with adaptive filtering.

For further information, see the PNG specification (RFC 2083),
<http://www.w3.org/pub/WWW/TR>.
.TP
.B "-scene \fIvalue\fP"
image scene number.
.TP
.B "-size \fI<width>x<height>{+offset}"
width and height of the image.

Use this option to specify the width and height of raw images whose
dimensions are unknown such as \fBGRAY\fP, \fBRGB\fP, or \fBCMYK\fP.
In addition to width and height, use \fB-size\fP to skip any header
information in the image or  tell the number of colors in a \fBMAP\fP
image file, (e.g. -size 640x512+256).
.TP
.B "-stegano \fIoffset\fP"
hide watermark within an image.
<p>
Use an offset to start the image hiding some number of pixels from the
beginning of the image.  Note this offset and the image size.  You will
need this information to recover the steganographic image
(e.g. display -size 320x256+35 stegano:image.png).
.TP
.B "-stereo"
composite two image to create a stereo anaglyph.

The left side of the stereo pair is saved as the red channel of the output
image.  The right sife is saved as the green channel.  Red-blue stereo
glasses are required to properly view the stereo image.
.TP
.B "-tile"
repeat composite operation across image.
.TP
.B "-treedepth \fIvalue\fP"
Normally, this integer value is zero or one.  A zero or one tells
\fBcomposite\fP to choose a optimal tree depth for the color reduction
algorithm.

An optimal depth generally allows the best representation of the source
image with the fastest computational speed and the least amount of
memory.  However, the default depth is inappropriate for some images.
To assure the best representation, try values between 2 and 8 for this
parameter.  Refer to \fBquantize(9)\fP for more details.

The \fB-colors\fP option is required for this option to take effect.
.TP
.B "-type \fItype\fP"
set the image type: 
\fBBilevel\fP, \fBGrayscale\fP, \fBPalette\fP,
\fBPaletteMatte\fP, \fBTrueColor\fP, \fBTrueColorMatte\fP,
\fBColorSeparation\fP, or \fBColorSeparationMatte\fP.
.TP
.B -verbose
print detailed information about the image.

This information is printed: image scene number;  image name;  composited
image name;  image size;  the image class (\fIDirectClass\fP or
\fIPseudoClass\fP);  the total number of unique colors;  and the number
of seconds to read and composite the image.
.PP
Options are processed in command line order.
Any option you specify on the command line remains in effect until it is
explicitly changed by specifying the option again with a different effect.

By default, the image format is determined by its magic number. To
specify a particular image format, precede the filename with an image
format name and a colon (i.e. ps:image) or specify the image type as
the filename suffix (i.e. image.ps).  See \fBconvert(1)\fP for a list
of valid image formats.

When you specify \fBX\fP as your image type, the filename has special
meaning.  It specifies an X window by id, name, or \fBroot\fP.  If no
filename is specified, the window is selected by clicking the mouse in
the desired window.

Specify \fIimage\fP as \fI-\fP for standard input,
\fIcomposited\fP as \fI-\fP for standard output.  If \fIimage\fP
has the extension \fB.Z\fP or \fB.gz\fP, the file is uncompressed with
\fBuncompress\fP or \fBgunzip\fP respectively.  If \fIcomposited\fP
has the extension \fB.Z\fP or \fB.gz\fP, the file size is compressed
using with \fBcompress\fP or \fBgzip\fP respectively.  Finally, precede
the image file name with \fI|\fP to pipe to or from a system command.

Use an optional index enclosed in brackets after a file name to specify
a desired subimage of a multi-resolution image format like Photo CD
(e.g. img0001.pcd[4]) or a range for MPEG images (e.g. video.mpg[50-75]).
A subimage specification can be disjoint (e.g. image.tiff[2,7,4]).
For raw images, specify a subimage with a geometry (e.g.  -size 640x512
image.rgb[320x256+50+50]).

The optional \fBmask\fP can be used to provide matte information for
\fBcomposite\fP when it has none or if you want a different mask.
A mask image is typically grayscale and the same size as
\fBcomposite\fP.  If the image is not grayscale, it is converted to
grayscale and the resulting intensities are used as matte information.

If \fIcomposited\fP already exists, you will be prompted as to whether
it should be overwritten.
.SH ENVIRONMENT

.TP
.B display
To get the default host, display number, and screen.
.SH SEE ALSO
.B
display(1), animate(1), import(1), montage(1), mogrify(1), convert(1), xtp(1)
.SH COPYRIGHT
Copyright (C) 2001 ImageMagick Studio, a non-profit organization dedicated
to making software imaging solutions freely available.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files ("ImageMagick"),
to deal in ImageMagick without restriction, including without limitation
the rights to use, copy, modify, merge, publish, distribute, sublicense,
and/or sell copies of ImageMagick, and to permit persons to whom the
ImageMagick is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of ImageMagick.

The software is provided "as is", without warranty of any kind, express or
implied, including but not limited to the warranties of merchantability,
fitness for a particular purpose and noninfringement.  In no event shall
ImageMagick Studio be liable for any claim, damages or other liability,
whether in an action of contract, tort or otherwise, arising from, out of
or in connection with ImageMagick or the use or other dealings in
ImageMagick.

Except as contained in this notice, the name of the ImageMagick Studio
shall not be used in advertising or otherwise to promote the sale, use or
other dealings in ImageMagick without prior written authorization from the
ImageMagick Studio.
.SH AUTHORS
John Cristy, E.I. du Pont De Nemours and Company Incorporated


NEWS
This file was last updated on April 1, 2009.

Significant changes associated with GraphicsMagick 1.1.15 (April 1, 2009)

  Bugs Fixed:

    o The MagickLibVersion define overflowed as of 1.1.10 since only
      one digit was reserved for the minor version.  Fixed now.

    o MSL: Fix for handling of geometry x,y values in association with
      gravity.

    o TIFF: Ping mode is now fast like it should have been.

    o Conjure subcommand was returning inverted status to the
      environment.

    o Benchmark subcommand was leaking memory for each iteration.

    o Magick++ Image color quantization error measurement was
      performing a wasteful operation.

    o GraphicsMagick.spec was not installing module .la files by default.

    o Bounding box query was failing rather than returning a bounding box
      matching the entire image if there was no subimage to bound.

    o Memory leak/use fixes.

  Security Fixes:

    o Exclude all virtual delegates and coders for pseudo-formats when
      deciding to use file extension as the format.

    o Fix assertion thrown for BMP and DIB files with negative image
      height values.

Significant changes associated with GraphicsMagick 1.1.14 (released June 5, 2008)

  Bugs Fixed:

    o Noise generation was not working correctly for anything but a Q8 build.

    o Poisson noise generator was extremely slow.  Now it is reasonably fast.

    o PDF reader now computes bounding box for rotated PDFs correctly.

  Security Fixes:

    o Cineon reader: Fixed crash with broken file from Sami Liedes.

    o PICT reader: Fixed crash with broken files from Sami Liedes.

    o XCF reader: Fixed crash with broken files from Sami Liedes.

  Security Issue Not Fixed:

    o JP2/JPC reader: Several JPEG-2000 files from Sami Liedes broken collection
      cause crashes in Jasper.  This is not GraphicsMagick's fault.

Significant changes associated with GraphicsMagick 1.1.13 (released May 28, 2008)

  Bugs Fixed:

    o Documentation is now installed according to the conventions established by
      the configure script.  This results in documentation being installed under
      /usr/local/share/doc/GraphicsMagick by default.

    o HWB colorspace now works correctly.

    o Composition with CopyOpacity now produces an image with
      transparency (as expected).

    o Composition now preserves the canvas colorspace.

    o Composition with a displacement map (-displace) no longer leaks an image.

    o Alpha composition now works as expected when both pixels involved
      include transparency.

    o -gamma multiple channel syntax now works as documented.

    o Now compiles with Visual C++ 2008.

Significant changes associated with GraphicsMagick 1.1.12 (released April 28, 2008)

  Security Fixes:

    o Do not access X11 or invoke convenience or stealth delegate
      programs based on the file extension. In particular, these file
      extensions are rejected for consideration as a format specifier:
      'autotrace', 'browse', 'dcraw', 'edit', 'gs-color',
      'gs-color+alpha', 'gs-gray', 'gs-mono', 'launch', 'mpeg-encode',
      'print', 'scan', 'show', 'win', 'xc', and 'x'.

  Bugs Fixed:

    o magick/effect.c: Should now compile for ARM CPU.

    o TIFF: Don't request Kodak private tags since these cause some
      versions of libtiff to misbehave.

    o When performing string expansion of image attribute identifiers, skip
      those which require access to image pixels if image pixels are not
      present.

    o CropImageToHBITMAP(), ImageToHBITMAP(): Fix leak of bitmap handle.

Significant changes associated with GraphicsMagick 1.1.11 (released February 23, 2008)

  Bugs Fixed:

    o BMP: Support large files.

    o DIB: Support large files.

    o PNG: Fix depth handling with 16-bit PNG files in the Q8 build.

    o SUN: Properly report image depth.

    o TIFF: Endian option (-endian) now controls TIFF byte endian order
      rather than bit fill order.

    o DCM, DIB, XBM, XCF, XWD: Eliminate integer overflow vulnerability
      (IDefense 09.19.07).

    o HSL colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o HWB colorspace transform: Avoid optimization bug noticed on Opteron
      with GCC.

    o RGBTransformImage()/TransformRGBImage(): Was using HWB colorspace
      when HSL was requested.

    o Successfully reads files with names like 'file[123]'.

    o 'gm display': No longer rely on isatty() to determine if input is
      from a pipe (use 'gm display -' to display an image read from a pipe).

  Feature Improvements:

    o 'identify +ping' forces the pixels to be read (similar to GM 1.2).

    o 'gm -version' now indicates if build supports "Large Memory" (i.e. 64-bit).

    o TIFF: Use '-define tiff:fill-order={msb2lsb|lsb2msb}' to control
      TIFF bit fill order.


  Performance Improvements:

    o No longer bogs down if a directory contains hundreds of thousands of
      files and the filename looks like a wildcard specification.


Significant changes associated with GraphicsMagick 1.1.10 (released September 19, 2007)

  Bugs Fixed:

    o Image rotate by -90/270 degrees was producing wrong output.

Significant changes associated with GraphicsMagick 1.1.9 (released September 16, 2007)

  Bugs Fixed:

    o In mogrify command, don't remove file name based on random junk in
      memory.
    o Fixed memory leak when reading MPC files.
    o Fixed crash when writing MIFF format and depth is not expected 8/16/32/.
    o Find latest Ghostscript under Windows which identifies itself as
      "GPL Ghostscript".
    o In mogrify command, don't leak memory in the case where the image
      file contains multiple frames.
    o Fixed crash in PNG and JPEG coders when the image to be written is
      part of an image list.
    o PNG reader errors are now properly reported to the user.
    o TIFF output can now be written to a pipe or other non-seekable
      destination.

PerlMagick/demo/composite.pl
#!/usr/local/bin/perl

use Graphics::Magick;

#$font = '-adobe-helvetica-medium-r-normal--25-180-100-100-p-130-iso8729-1';
#$font = 'Times';
$font = 'Generic.ttf';

$image = Graphics::Magick->new();
$smile = Graphics::Magick->new();
$smile->Read('smile.gif');
$smile->Set(background=>'none');
$x = 100;
$y = 100;
for ($angle=0; $angle < 360; $angle+=30)
{
  my ($thumbnail);

  print "angle $angle\n";
  $thumbnail=Graphics::Magick->new(size=>"600x600",pointsize=>24,font=>$font,
    fill=>'black');
  $thumbnail->Read("xc:white");
  $thumbnail->Draw(primitive=>'line',points=>"300,100 300,500",stroke=>'#600');
  $thumbnail->Draw(primitive=>'line',points=>"100,300 500,300",stroke=>'#600');
  $thumbnail->Draw(primitive=>'rectangle',points=>"100,100 500,500",
    fill=>'none',stroke=>'#600');
  $thumbnail->Composite(image=>$smile,gravity=>"NorthWest",x=>$x,y=>$y,
    rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"North",y=>$y,rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"NorthEast",x=>$x,y=>$y,
    rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"West",x=>$x,rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"Center",rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"East",x=>$x,rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"SouthWest",x=>$x,y=>$y,
    rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"South",y=>$y,rotate=>$angle);
  $thumbnail->Composite(image=>$smile,gravity=>"SouthEast",x=>$x,y=>$y,
    rotate=>$angle);
  push(@$image,$thumbnail);
}
$image->Set(delay=>20);
$image->Write("composite.miff");
$image->Animate();
